sdk/core/Azure.Core/src/ResourceType.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Globalization;
+using System.Linq;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Structure representing a resource type.
+    /// </summary>
+    public readonly struct ResourceType : IEquatable<ResourceType>
+    {
+        private const char Separator = '/';
+        private readonly string _stringValue;
+
+        /// <summary>
+        /// The resource type for the root of the resource hierarchy.
+        /// </summary>
+        public static ResourceType Root { get; } = new ResourceType(string.Empty, string.Empty);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ResourceType"/> class.
+        /// </summary>
+        /// <param name="resourceType"> The resource type string to convert. </param>
+        public ResourceType(string resourceType)
+        {
+            Argument.AssertNotNullOrWhiteSpace(resourceType, nameof(resourceType));
+
+            int index = resourceType.IndexOf(Separator);
+            if (index == -1 || resourceType.Length <3)
+                throw new ArgumentOutOfRangeException(nameof(resourceType));
+
+            _stringValue = resourceType;
+            Namespace = resourceType.Substring(0, index);
+            Type = resourceType.Substring(index + 1);
+        }
+
+        /// <summary>
+        /// Create a resource type given the namespace and typename components.
+        /// </summary>
+        /// <param name="providerNamespace"></param>
+        /// <param name="name"></param>
+        internal ResourceType(string providerNamespace, string name)
+        {
+            Namespace = providerNamespace;
+            Type = name;
+            _stringValue = $"{Namespace}{Separator}{Type}";
+        }
+
+        internal ResourceType AppendChild(string childType)
+        {
+            return new ResourceType($"{_stringValue}{Separator}{childType}");
+        }
+
+        /// <summary>
+        /// Gets the last resource type name.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
```

Comment:
Curious why we're adding a public method that's not visible to customers?  It looks like it's used in Azure.ResourceManager - would it make sense to add an extension method there?

Reply:
its used in each of the RPs to create their IDs from their parents.  It will also be used in powershell, but we don't necessarily think customers will use this directly